File: 4.sort/1051.HeightChecker/main.go
package main

import "sort"

// https://leetcode.com/problems/height-checker/
// TC: O(nlogn), SC: O(n)
func heightChecker1(heights []int) int {
	tempHeights := make([]int, len(heights))
	copy(tempHeights, heights)
	sort.Ints(tempHeights)

	result := 0
	for i := range heights {
		if heights[i] != tempHeights[i] {
			result++
		}
	}

	return result
}

func heightChecker2(heights []int) int {
	freq := make([]int, 101)
	for _, h := range heights {
		freq[h]++
	}

	diff := 0
	// current index of heights
	i := 0
	for h, f := range freq {
		for f > 0 {
			// expected[i] = h
			if heights[i] != h {
				diff++
			}
			i++
			f--
		}
	}

	return diff
}

File: 4.sort/1051.HeightChecker/main.py
from typing import List

# https://leetcode.com/problems/height-checker/
# counting sort
# [2,1,1,4]
# [0,2,1,0,1]
# TC: O(101) ~ O(1), SC: O(101) ~ O(1)
class Solution:
    def heightChecker(self, heights: List[int]) -> int:
        counts = [0] * 101
        for h in heights:
            counts[h] += 1
        curIdx = 0
        diff = 0
        for i in range(len(counts)):
            while counts[i] > 0:
                if heights[curIdx] != i:
                    diff += 1
                counts[i] -= 1
                curIdx += 1
        return diff

# O(nlogn), SC: O(n)
class Solution:
    def heightChecker(self, heights: List[int]) -> int:
        sortedHeights = sorted(heights)
        diff = 0
        for i in range(len(heights)):
            if sortedHeights[i] != heights[i]:
                diff += 1
        return diff
